File: 0084 largest-rectangle-in-histogram/solution.cs
public class Solution {
    public int LargestRectangleArea(int[] h) {
        Stack<int> st = new();
        int max=0;
        for(int i=0;i<=h.Length;i++) {
            int he=(i == h.Length ? 0 : h[i]);
            while(st.Count > 0 && he<h[st.Peek()]){
                int height=h[st.Pop()];
                int width = (st.Count == 0 ? i : i - st.Peek() - 1);
                max=Math.Max(max,width*height);
            }
            st.Push(i);
        }
        return max;
    }
}

File: 0084 largest-rectangle-in-histogram/solution.rs
impl Solution {
    pub fn largest_rectangle_area(mut heights: Vec<i32>) -> i32 {
        heights.push(0); 
        let mut stack: Vec<usize> = Vec::new();
        let mut max_area = 0;
        for i in 0..heights.len() {
            let current_height = heights[i];
            while let Some(&last_idx) = stack.last() {
                if heights[last_idx] >= current_height {
                    let h = heights[stack.pop().unwrap()]; 
                    
                    let width = if stack.is_empty() {
                        i as i32 
                    } else {
                        (i - stack.last().unwrap() - 1) as i32
                    };

                    max_area = max_area.max(h * width);
                } else {
                    break;
                }
            }
            stack.push(i);
        }
        max_area
    }
}
